--- linux-3.2.40.orig/include/linux/serial_core.h	2013-03-05 19:24:30.000000000 -0800
+++ linux-3.2.40/include/linux/serial_core.h	2013-04-13 23:50:14.000000000 -0700
@@ -320,6 +320,12 @@
 #define UPIO_MEM32		(3)
 #define UPIO_AU			(4)			/* Au1x00 type IO */
 #define UPIO_TSI		(5)			/* Tsi108/109 type IO */
+
+#if defined(CONFIG_ARCH_ARMADA370) || defined(CONFIG_ARCH_ARMADA_XP)
+#define UPIO_DWAPB		(6)
+#define UPIO_DWAPB32		(7)
+#endif
+
 #define UPIO_RM9000		(6)			/* RM9000 type IO */
 
 	unsigned int		read_status_mask;	/* driver specific */
