--- linux-3.2.40.orig/arch/arm/plat-armada/mv_hal/qd-dsdt/sample/Trailer/trailer.c	1969-12-31 16:00:00.000000000 -0800
+++ linux-3.2.40/arch/arm/plat-armada/mv_hal/qd-dsdt/sample/Trailer/trailer.c	2013-04-13 23:49:38.000000000 -0700
@@ -0,0 +1,50 @@
+#include <Copyright.h>
+/********************************************************************************
+* trailer.c
+*
+* DESCRIPTION:
+*		This sample shows how to enable/disable CPU port's ingress and egress
+*		Trailer mode.
+*
+* DEPENDENCIES:   NONE.
+*
+* FILE REVISION NUMBER:
+*
+*******************************************************************************/
+
+#include "msSample.h"
+
+GT_STATUS sampleCPUTrailerEnable(GT_QD_DEV *dev, GT_BOOL en)
+{
+	GT_STATUS status;
+	GT_INGRESS_MODE	inMode;
+
+	if (en)	/* Enable Trailer Mode */
+	{
+		inMode = GT_TRAILER_INGRESS;
+	}
+	else
+	{
+		inMode = GT_UNMODIFY_INGRESS;
+	}
+
+	/*
+	 *	Enable CPU's Ingress Trailer
+	*/
+	if((status = gprtSetIngressMode(dev,dev->cpuPortNum, inMode)) != GT_OK)
+	{
+		MSG_PRINT(("gprtSetIngressMode return Failed\n"));
+		return status;
+	}
+
+	/*
+	 *	Enable CPU's Egress Trailer
+	*/
+	if((status = gprtSetTrailerMode(dev,dev->cpuPortNum, en)) != GT_OK)
+	{
+		MSG_PRINT(("gprtSetTrailerMode return Failed\n"));
+		return status;
+	}
+
+	return GT_OK;
+}
