--- linux-3.2.40.orig/arch/arm/plat-armada/mv_drivers_lsp/mv_audio_soc/mv88fx-i2s.h	1969-12-31 16:00:00.000000000 -0800
+++ linux-3.2.40/arch/arm/plat-armada/mv_drivers_lsp/mv_audio_soc/mv88fx-i2s.h	2013-04-12 17:28:36.000000000 -0700
@@ -0,0 +1,20 @@
+/*
+ *
+ *	Marvell Orion Alsa SOC Sound driver
+ *
+ *	Author: Yuval Elmaliah
+ *	Copyright (C) 2008 Marvell Ltd.
+ *
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ *
+ */
+
+#ifndef _MV88FX_I2S_H
+#define _MV88FX_I2S_H
+
+extern struct snd_soc_dai mv88fx_i2s_dai;
+
+#endif
