--- linux-3.2.40.orig/drivers/net/phy/marvell.c	2013-03-05 19:24:30.000000000 -0800
+++ linux-3.2.40/drivers/net/phy/marvell.c	2013-04-13 23:50:00.000000000 -0700
@@ -271,7 +271,7 @@
 
 static int m88e1121_config_aneg(struct phy_device *phydev)
 {
-	int err, oldpage, mscr;
+	int err, oldpage, mscr, val;
 
 	oldpage = phy_read(phydev, MII_MARVELL_PHY_PAGE);
 
@@ -315,7 +315,11 @@
 	oldpage = phy_read(phydev, MII_MARVELL_PHY_PAGE);
 
 	phy_write(phydev, MII_MARVELL_PHY_PAGE, MII_88E1121_PHY_LED_PAGE);
-	phy_write(phydev, MII_88E1121_PHY_LED_CTRL, MII_88E1121_PHY_LED_DEF);
+	/* Fixes power-down issue on RD-88F6282-A board by bypassing WoL logic */
+	val = phy_read(phydev, MII_88E1121_PHY_LED_CTRL);
+	val &= ~0xfff;
+	val |= 0x813;
+	phy_write(phydev, MII_88E1121_PHY_LED_CTRL, val);
 	phy_write(phydev, MII_MARVELL_PHY_PAGE, oldpage);
 
 	err = genphy_config_aneg(phydev);
