--- linux-3.2.40.orig/include/linux/skbuff.h	2013-03-05 19:24:30.000000000 -0800
+++ linux-3.2.40/include/linux/skbuff.h	2013-04-13 23:50:14.000000000 -0700
@@ -31,6 +31,10 @@
 #include <linux/hrtimer.h>
 #include <linux/dma-mapping.h>
 
+#if defined(CONFIG_NET_SKB_HEADROOM)
+# define NET_SKB_PAD  CONFIG_NET_SKB_HEADROOM
+#endif
+
 /* Don't change this without changing skb_csum_unnecessary! */
 #define CHECKSUM_NONE 0
 #define CHECKSUM_UNNECESSARY 1
@@ -415,6 +419,10 @@
 	__be16			protocol;
 
 	void			(*destructor)(struct sk_buff *skb);
+#ifdef CONFIG_NET_SKB_RECYCLE
+	int			(*skb_recycle) (struct sk_buff *skb);
+	void			*hw_cookie;
+#endif /* CONFIG_NET_SKB_RECYCLE */
 #if defined(CONFIG_NF_CONNTRACK) || defined(CONFIG_NF_CONNTRACK_MODULE)
 	struct nf_conntrack	*nfct;
 #endif
@@ -2085,6 +2093,9 @@
 extern int	       skb_copy_datagram_iovec(const struct sk_buff *from,
 					       int offset, struct iovec *to,
 					       int size);
+extern int		skb_copy_datagram_to_kernel_iovec(const struct sk_buff *from,
+					       int offset, struct iovec *to,
+					       int size);
 extern int	       skb_copy_and_csum_datagram_iovec(struct sk_buff *skb,
 							int hlen,
 							struct iovec *iov);
@@ -2525,7 +2536,7 @@
 	if (skb_end_pointer(skb) - skb->head < skb_size)
 		return false;
 
-	if (skb_shared(skb) || skb_cloned(skb))
+	if (skb_shared(skb) || skb_cloned(skb) || skb_has_frag_list(skb))
 		return false;
 
 	return true;
