--- linux-3.2.40.orig/arch/arm/mach-armadaxp/mpp.h	1969-12-31 16:00:00.000000000 -0800
+++ linux-3.2.40/arch/arm/mach-armadaxp/mpp.h	2013-04-12 17:28:36.000000000 -0700
@@ -0,0 +1,34 @@
+#ifndef __ARCH_DOVE_MPP_H
+#define __ARCH_DOVE_MPP_H
+
+enum aurora_mpp_type {
+	/*
+	 * This MPP is unused.
+	 */
+	MPP_UNUSED,
+
+	/*
+	 * This MPP pin is used as a generic GPIO pin.
+	 */
+	MPP_GPIO,
+
+        /*
+         * This MPP is used as a SATA activity LED.
+         */
+        MPP_SATA_LED,
+        /*
+         * This MPP is used as a functional pad.
+         */
+        MPP_FUNCTIONAL,
+
+};
+
+struct aurora_mpp_mode {
+	int			mpp;
+	enum aurora_mpp_type	type;
+};
+
+void aurora_mpp_conf(struct aurora_mpp_mode *mode);
+
+
+#endif
