--- linux-3.2.40.orig/arch/arm/plat-armada/mv_drivers_lsp/mv_network/mv_ethernet/Makefile	1969-12-31 16:00:00.000000000 -0800
+++ linux-3.2.40/arch/arm/plat-armada/mv_drivers_lsp/mv_network/mv_ethernet/Makefile	2013-04-12 17:28:38.000000000 -0700
@@ -0,0 +1,13 @@
+#
+# Makefile for the Marvell Gigabit Ethernet driver
+#
+ifneq ($(MACHINE),)
+include $(srctree)/$(MACHINE)/config/mvRules.mk
+endif
+
+obj-$(CONFIG_MV_ETHERNET) += mv_netdev.o mv_ethernet.o
+obj-$(CONFIG_MV_ETH_PROC) += mv_eth_proc.o
+obj-$(CONFIG_MV_GATEWAY) +=  mv_gateway.o
+obj-$(CONFIG_MV_GTW_IGMP) += mv_gtw_igmp.o
+obj-$(CONFIG_MV_ETH_TOOL) += mv_eth_tool.o
+
