--- linux-3.2.40.orig/mm/mmap.c	2013-03-05 19:24:30.000000000 -0800
+++ linux-3.2.40/mm/mmap.c	2013-04-13 23:50:16.000000000 -0700
@@ -1586,7 +1586,15 @@
 				vma_tmp = rb_entry(rb_node,
 						struct vm_area_struct, vm_rb);
 
+#ifdef CONFIG_MV_SUPPORT_64KB_PAGE_SIZE
+				/* Take into account a wrap-around of the
+				** vm_end field to 0x0. e.g. vm_start =
+				** 0xFFFF0000 size PAGE_SIZE.
+				*/
+				if ((vma_tmp->vm_end - 1) >= addr) {
+#else
 				if (vma_tmp->vm_end > addr) {
+#endif
 					vma = vma_tmp;
 					if (vma_tmp->vm_start <= addr)
 						break;
